File: frontend/src/api/quiz.ts
import request from './request'

export interface MessageItem {
  sequence: number
  role: string
  content_type: string
  content: string
  sender_name: string
  original_time?: string | null
}

export interface UploadPreviewData {
  preview_id: string
  source_file: string
  patient_name?: string | null
  counselor_name?: string | null
  message_count: number
  messages: MessageItem[]
}

export interface ConfirmPayload {
  preview_id: string
  title: string
  scope: 'common' | 'hospital' | 'department'
  hospital_id?: number
  department_id?: number
  chat_type: 'active' | 'passive'
  category?: string
  difficulty: number
  tags?: string
  description?: string
}

export interface UpdateQuizPayload {
  title: string
  scope: 'common' | 'hospital' | 'department'
  hospital_id?: number
  department_id?: number
  chat_type: 'active' | 'passive'
  category?: string
  difficulty: number
  tags?: string
  description?: string
  patient_name?: string
  counselor_name?: string
}

export interface QuizMetaOptionItem {
  name: string
  count: number
}

export interface QuizMetaOptionsData {
  categories: QuizMetaOptionItem[]
  tags: QuizMetaOptionItem[]
}

export interface BatchQuizMetaUpdatePayload {
  quiz_ids?: number[]
  keyword?: string
  scope?: 'common' | 'hospital' | 'department'
  hospital_id?: number
  department_id?: number
  chat_type?: 'active' | 'passive'
  set_category?: string
  clear_category?: boolean
  add_tags?: string[]
  remove_tags?: string[]
  replace_tags?: string[]
  clear_tags?: boolean
}

export interface QuizMetaOperateFilters {
  scope?: 'common' | 'hospital' | 'department'
  hospital_id?: number
  department_id?: number
  chat_type?: 'active' | 'passive'
}

export interface BatchReparsePayload {
  scope: 'common' | 'hospital' | 'department'
  hospital_id?: number
  department_id?: number
  chat_type?: 'active' | 'passive'
  limit?: number
  only_legacy_or_empty_hash?: boolean
}

export interface BatchReparseResult {
  matched: number
  processed: number
  updated: number
  skipped: number
  failed: number
  detail: Array<Record<string, any>>
}

export interface BatchReparseEstimateResult {
  matched: number
  limit: number
  only_legacy_or_empty_hash: boolean
}

export interface QuizListItem {
  id: number
  title: string
  scope: 'common' | 'hospital' | 'department'
  hospital_id?: number | null
  hospital_name?: string | null
  department_id?: number | null
  department_name?: string | null
  chat_type: 'active' | 'passive'
  category?: string | null
  difficulty: number
  tags?: string | null
  patient_name?: string | null
  counselor_name?: string | null
  message_count: number
  is_active: boolean
  is_deleted?: boolean
  created_at: string
}

export interface QuizDetailData extends QuizListItem {
  description?: string | null
  source_file?: string | null
  versions?: Array<{
    id: number
    version_no: number
    source_file?: string | null
    message_count: number
    created_by?: number | null
    created_at: string
  }>
  messages: MessageItem[]
}

interface ApiResponse<T> {
  code: number
  message: string
  data: T
}

export const uploadQuizHtml = async (file: File) => {
  const formData = new FormData()
  formData.append('file', file)
  const res = await request.post<ApiResponse<UploadPreviewData>>('/quizzes/upload', formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
  return res.data.data
}

export const confirmQuizImport = async (payload: ConfirmPayload) => {
  const res = await request.post<ApiResponse<{ quiz_id: number; title: string; message_count: number }>>(
    '/quizzes/confirm',
    payload,
  )
  return res.data.data
}

export const confirmQuizNewVersion = async (quizId: number, payload: ConfirmPayload) => {
  const res = await request.post<ApiResponse<{ quiz_id: number; title: string; message_count: number }>>(
    `/quizzes/${quizId}/versions/confirm`,
    payload,
  )
  return res.data.data
}

export const batchReparseQuizzes = async (payload: BatchReparsePayload) => {
  const res = await request.post<ApiResponse<BatchReparseResult>>('/quizzes/reparse/batch', payload)
  return res.data.data
}

export const estimateBatchReparseQuizzes = async (payload: BatchReparsePayload) => {
  const res = await request.post<ApiResponse<BatchReparseEstimateResult>>('/quizzes/reparse/batch/estimate', payload)
  return res.data.data
}

export const getQuizList = async (params: {
  page: number
  page_size: number
  keyword?: string
  scope?: 'common' | 'hospital' | 'department'
  hospital_id?: number
  department_id?: number
  chat_type?: 'active' | 'passive'
  deleted_only?: boolean
  include_deleted?: boolean
}) => {
  const res = await request.get<
    ApiResponse<{ items: QuizListItem[]; total: number; page: number; page_size: number }>
  >('/quizzes', { params })
  return res.data.data
}

export const getQuizDetail = async (id: number) => {
  const res = await request.get<ApiResponse<QuizDetailData>>(`/quizzes/${id}`)
  return res.data.data
}

export const softDeleteQuiz = async (id: number) => {
  const res = await request.delete<ApiResponse<null>>(`/quizzes/${id}`)
  return res.data
}

export const hardDeleteQuiz = async (id: number) => {
  const res = await request.delete<ApiResponse<null>>(`/quizzes/${id}`, {
    params: { hard: true },
  })
  return res.data
}

export const restoreQuiz = async (id: number) => {
  const res = await request.post<ApiResponse<null>>(`/quizzes/${id}/restore`)
  return res.data
}

export const updateQuiz = async (id: number, payload: UpdateQuizPayload) => {
  const res = await request.put<ApiResponse<{ quiz_id: number; title: string }>>(`/quizzes/${id}`, payload)
  return res.data.data
}

export const getQuizMetaOptions = async (params?: {
  scope?: 'common' | 'hospital' | 'department'
  hospital_id?: number
  department_id?: number
  chat_type?: 'active' | 'passive'
}) => {
  const res = await request.get<ApiResponse<QuizMetaOptionsData>>('/quizzes/meta/options', { params })
  return res.data.data
}

export const batchUpdateQuizMeta = async (payload: BatchQuizMetaUpdatePayload) => {
  const res = await request.post<ApiResponse<{ matched: number; updated: number }>>('/quizzes/batch/meta/update', payload)
  return res.data.data
}

export const renameQuizCategory = async (
  payload: QuizMetaOperateFilters & { old_name: string; new_name: string },
) => {
  const res = await request.post<ApiResponse<{ matched: number; updated: number }>>('/quizzes/meta/category/rename', payload)
  return res.data.data
}

export const mergeQuizCategories = async (
  payload: QuizMetaOperateFilters & { source_names: string[]; target_name: string },
) => {
  const res = await request.post<ApiResponse<{ matched: number; updated: number }>>('/quizzes/meta/category/merge', payload)
  return res.data.data
}

export const deleteQuizCategories = async (
  payload: QuizMetaOperateFilters & { names: string[] },
) => {
  const res = await request.post<ApiResponse<{ matched: number; updated: number }>>('/quizzes/meta/category/delete', payload)
  return res.data.data
}

export const renameQuizTag = async (
  payload: QuizMetaOperateFilters & { old_name: string; new_name: string },
) => {
  const res = await request.post<ApiResponse<{ matched: number; updated: number }>>('/quizzes/meta/tag/rename', payload)
  return res.data.data
}

export const mergeQuizTags = async (
  payload: QuizMetaOperateFilters & { source_names: string[]; target_name: string },
) => {
  const res = await request.post<ApiResponse<{ matched: number; updated: number }>>('/quizzes/meta/tag/merge', payload)
  return res.data.data
}

export const deleteQuizTags = async (
  payload: QuizMetaOperateFilters & { names: string[] },
) => {
  const res = await request.post<ApiResponse<{ matched: number; updated: number }>>('/quizzes/meta/tag/delete', payload)
  return res.data.data
}

File: backend/app/schemas/quiz.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

ChatType = Literal["active", "passive"]


class MessageItem(BaseModel):
    sequence: int
    role: str
    content_type: str
    content: str
    sender_name: str
    original_time: str | None = None


class ImportQualityReport(BaseModel):
    total_messages: int = 0
    patient_count: int = 0
    counselor_count: int = 0
    text_count: int = 0
    image_count: int = 0
    audio_count: int = 0
    skipped_empty: int = 0
    role_alternation_rate: float = 0.0
    unique_patient_names: list[str] = []
    unique_counselor_names: list[str] = []
    name_collision: bool = False
    quality_score: float = 0.0
    warnings: list[str] = []


class UploadPreviewData(BaseModel):
    preview_id: str
    source_file: str
    patient_name: str | None = None
    counselor_name: str | None = None
    message_count: int
    messages: list[MessageItem]
    quality_report: ImportQualityReport | None = None


class ConfirmImportRequest(BaseModel):
    preview_id: str
    title: str = Field(min_length=1, max_length=200)
    scope: Literal["common", "hospital", "department"] = "hospital"
    hospital_id: int | None = None
    department_id: int | None = None
    chat_type: ChatType
    category: str | None = Field(default=None, max_length=50)
    difficulty: int = Field(default=1, ge=1, le=5)
    tags: str | None = Field(default=None, max_length=500)
    description: str | None = None


class BatchReparseRequest(BaseModel):
    scope: Literal["common", "hospital", "department"] = "hospital"
    hospital_id: int | None = None
    department_id: int | None = None
    chat_type: ChatType | None = None
    limit: int = Field(default=500, ge=1, le=2000)
    only_legacy_or_empty_hash: bool = False


class QuizVersionItem(BaseModel):
    id: int
    version_no: int
    source_file: str | None = None
    message_count: int
    created_by: int | None = None
    created_at: datetime


class ConfirmImportData(BaseModel):
    quiz_id: int
    title: str
    message_count: int


class QuizUpdateRequest(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    scope: Literal["common", "hospital", "department"] = "hospital"
    hospital_id: int | None = None
    department_id: int | None = None
    chat_type: ChatType
    category: str | None = Field(default=None, max_length=50)
    difficulty: int = Field(default=1, ge=1, le=5)
    tags: str | None = Field(default=None, max_length=500)
    description: str | None = None
    patient_name: str | None = Field(default=None, max_length=100)
    counselor_name: str | None = Field(default=None, max_length=100)


class QuizUpdateData(BaseModel):
    quiz_id: int
    title: str
    scope: Literal["common", "hospital", "department"]
    hospital_id: int | None = None
    department_id: int | None = None
    chat_type: ChatType
    category: str | None = None
    difficulty: int
    tags: str | None = None
    description: str | None = None
    patient_name: str | None = None
    counselor_name: str | None = None


class QuizMetaOptionItem(BaseModel):
    name: str
    count: int


class QuizMetaOptionsData(BaseModel):
    categories: list[QuizMetaOptionItem] = []
    tags: list[QuizMetaOptionItem] = []


class BatchQuizMetaUpdateRequest(BaseModel):
    quiz_ids: list[int] = []
    keyword: str | None = None
    scope: Literal["common", "hospital", "department"] | None = None
    hospital_id: int | None = None
    department_id: int | None = None
    chat_type: ChatType | None = None
    set_category: str | None = Field(default=None, max_length=50)
    clear_category: bool = False
    add_tags: list[str] = []
    remove_tags: list[str] = []
    replace_tags: list[str] | None = None
    clear_tags: bool = False


class BatchQuizMetaUpdateData(BaseModel):
    matched: int
    updated: int


class QuizMetaOperateFilter(BaseModel):
    scope: Literal["common", "hospital", "department"] | None = None
    hospital_id: int | None = None
    department_id: int | None = None
    chat_type: ChatType | None = None


class CategoryRenameRequest(QuizMetaOperateFilter):
    old_name: str = Field(min_length=1, max_length=50)
    new_name: str = Field(min_length=1, max_length=50)


class CategoryMergeRequest(QuizMetaOperateFilter):
    source_names: list[str] = []
    target_name: str = Field(min_length=1, max_length=50)


class CategoryDeleteRequest(QuizMetaOperateFilter):
    names: list[str] = []


class TagRenameRequest(QuizMetaOperateFilter):
    old_name: str = Field(min_length=1, max_length=50)
    new_name: str = Field(min_length=1, max_length=50)


class TagMergeRequest(QuizMetaOperateFilter):
    source_names: list[str] = []
    target_name: str = Field(min_length=1, max_length=50)


class TagDeleteRequest(QuizMetaOperateFilter):
    names: list[str] = []


class QuizMetaOperateData(BaseModel):
    matched: int
    updated: int


class BatchReparseData(BaseModel):
    matched: int
    processed: int
    updated: int
    skipped: int
    failed: int
    detail: list[dict] = []


class BatchReparseEstimateData(BaseModel):
    matched: int
    limit: int
    only_legacy_or_empty_hash: bool


class QuizListItem(BaseModel):
    id: int
    title: str
    scope: Literal["common", "hospital", "department"]
    hospital_id: int | None = None
    hospital_name: str | None = None
    department_id: int | None = None
    department_name: str | None = None
    chat_type: ChatType
    category: str | None = None
    difficulty: int
    tags: str | None = None
    patient_name: str | None = None
    counselor_name: str | None = None
    message_count: int
    is_active: bool
    is_deleted: bool = False
    created_at: datetime


class QuizDetailData(BaseModel):
    id: int
    title: str
    scope: Literal["common", "hospital", "department"]
    hospital_id: int | None = None
    hospital_name: str | None = None
    department_id: int | None = None
    department_name: str | None = None
    chat_type: ChatType
    description: str | None = None
    category: str | None = None
    difficulty: int
    tags: str | None = None
    patient_name: str | None = None
    counselor_name: str | None = None
    message_count: int
    source_file: str | None = None
    is_active: bool
    is_deleted: bool = False
    created_at: datetime
    versions: list[QuizVersionItem] = []
    messages: list[MessageItem]


class PageData(BaseModel):
    items: list[QuizListItem]
    total: int
    page: int
    page_size: int


class ApiResponse(BaseModel):
    code: int
    message: str
    data: (
        UploadPreviewData
        | ConfirmImportData
        | QuizUpdateData
        | QuizMetaOptionsData
        | BatchQuizMetaUpdateData
        | QuizMetaOperateData
        | BatchReparseData
        | BatchReparseEstimateData
        | QuizDetailData
        | PageData
        | None
    ) = None
